lib/bandit/http1/adapter.ex

```diff
@@ -286,23 +286,55 @@ defmodule Bandit.HTTP1.Adapter do
   # Internal Reading
   ##################
 
-  @dialyzer {:no_improper_lists, read: 5}
-  defp read(socket, to_read, opts \\ [], bytes_read \\ 0, already_read \\ []) do
+  @spec read(
+          ThousandIsland.Socket.t(),
+          non_neg_integer(),
+          [read_length: non_neg_integer(), read_timeout: timeout()],
+          non_neg_integer(),
+          iodata()
+        ) ::
+          {:ok, iolist()}
+          | {:error, :closed | :timeout | :inet.posix()}
+  defp read(socket, to_read, opts \\ [], _bytes_read \\ 0, already_read \\ []) do
     read_size = min(to_read, Keyword.get(opts, :read_length, 1_000_000))
     read_timeout = Keyword.get(opts, :read_timeout)
+    # when to_read is 0, we still want to read at least 1 byte
+    # or everything which is already in the buffer of the socket
+    remaining_bytes = max(to_read, 1)
```

Comment:
Are you sure this won't end up with an off by one error in the case where we happen to get a read chunk that completely satisfies our desired length?

Reply:
Please ignore it. `defmodule ReadWith3 do` contains the latest and greats implementation actually.